runtime/caches/tiered.ts

```diff
@@ -0,0 +1,136 @@
+const inFuture = (maybeDate: string) => {
+  try {
+    return new Date(maybeDate) > new Date();
+  } catch {
+    return false;
+  }
+};
+
+/**
+ * Creates a tiered cache that combines multiple cache storages into a single cache storage.
+ * The tiered cache will prioritize the caches in the order they are provided.
+ * When a request is made to the tiered cache, it will first check if the request is available in any of the caches.
+ * If a match is found, the response will be returned and the caches will be updated based on the cache priority.
+ * If no match is found, the request will be fetched and stored in the caches based on the cache priority.
+ * @param tieredCaches The cache storages to be combined into the tiered cache.
+ * @returns The tiered cache storage.
+ */
+export function createTieredCache(
+  ...tieredCaches: (CacheStorage | undefined)[]
+): CacheStorage {
+  const openedCaches: Cache[] = [];
+  async function updateTieredCaches(
+    indexOfCachesToUpdate: number[],
+    request: RequestInfo | URL,
+    matched: Response,
+  ) {
+    const putPromises = indexOfCachesToUpdate.map((index) =>
+      openedCaches[index].put(request, matched.clone())
+    );
+    await Promise.all(putPromises);
+  }
+
+  const caches: CacheStorage = {
+    delete: (_cacheName: string): Promise<boolean> => {
+      throw new Error("Not Implemented");
+    },
+    has: (_cacheName: string): Promise<boolean> => {
+      throw new Error("Not Implemented");
+    },
+    keys: (): Promise<string[]> => {
+      throw new Error("Not Implemented");
+    },
+    match: (
+      _request: URL | RequestInfo,
+      _options?: MultiCacheQueryOptions | undefined,
+    ): Promise<Response | undefined> => {
+      throw new Error("Not Implemented");
+    },
+    open: async (cacheName: string): Promise<Cache> => {
+      await Promise.all(
+        tieredCaches.filter(Boolean).map(async (cache) => {
+          if (cache) {
+            openedCaches.push(await cache.open(cacheName));
+          }
+        }),
+      );
+      return Promise.resolve({
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/add) */
+        add: (_request: RequestInfo | URL): Promise<void> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/addAll) */
+        addAll: (_requests: RequestInfo[]): Promise<void> => {
+          throw new Error("Not Implemented");
+        },
+        /** [MDN Reference](https://developer.mozilla.org/docs/Web/API/Cache/delete) */
+        delete: async (
+          request: RequestInfo | URL,
+          options?: CacheQueryOptions,
+        ): Promise<boolean> => {
+          let deleteResponse;
+          for (const caches of openedCaches) {
+            deleteResponse = await caches.delete(request, options);
+          }
+          return deleteResponse ?? false;
```

Comment:
do in parallel.

Promise.all(openedCaches.map(cache => cache.delete(request, options)).then(results => results.every(r => r))